shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2860,3 +2925,50 @@ impl<'a> std::fmt::Display for FormatTopicName<'a> {
         }
     }
 }
+
+/// Partitions the given shotover nodes based on their rack placement.
+/// Returns a tuple of two vectors: (shotover_nodes_in_rack, shotover_nodes_out_of_rack).
+/// If rack is None, all shotover nodes are considered to be in the rack.
+fn partition_shotover_nodes_by_rack<'a>(
+    shotover_nodes: &[&'a ShotoverNode],
+    rack: &Option<&StrBytes>,
+) -> (Vec<&'a ShotoverNode>, Vec<&'a ShotoverNode>) {
+    shotover_nodes
+        .iter()
+        .partition(|shotover_node| rack.map(|rack| rack == &shotover_node.rack).unwrap_or(true))
+}
+
+/// Tries to deterministically select a shotover node based on a hash value.
+/// Prioritizes nodes in the rack if available, falls back to nodes out of the rack.
+fn select_shotover_node_by_hash<'a>(
```

Comment:
Two suggestions here:
1.  lets treat having at least one shotover node as an invariant and panic if that isnt true
2. we can use `.or_else` to avoid the is_empty checks.
```
fn select_shotover_node_by_hash<'a>(
    shotover_nodes_in_rack: Vec<&'a ShotoverNode>,
    shotover_nodes_out_of_rack: Vec<&'a ShotoverNode>,
    hash: usize,
) -> &'a ShotoverNode {
    shotover_nodes_in_rack
        .get(hash % shotover_nodes_in_rack.len())
        .cloned()
        .or_else(|| {
            shotover_nodes_out_of_rack
                .get(hash % shotover_nodes_out_of_rack.len())
                .cloned()
        })
        .expect("There must be at least one shotover node")
}
```

Reply:
I've changed it to a different version as `shotover_nodes_in_rack.get(hash % shotover_nodes_in_rack.len())` will panic if shotover_nodes_in_rack.len()=0